meta-lxatac-bsp: emmc-image: build smaller images (faster)

hnez opened the pull request:
This PR does three things:

- Revert #62 by @rohieb because it did not achieve quite what we wanted
- Remove the `partition end` workaround that was removed in #62 (but without changing other things like adding `fill=true`).
- Do not place a GPT header at the "end"¹ of the image.
- Use the existing rootfs ext4 instead of generating a new one from the `.tar`.

Things I am not too sure about:

- [x] Should we use some templating magic instead of writing `lxatac-core-image-base-lxatac.ext4` in the `genimage.config`?
- [x] Are all the dependencies automatically set up so that we can rely on `lxatac-core-image-base-lxatac.ext4` being present.
     The first version of this PR used `GENIMAGE_ROOTFS_IMAGE` to specify this dependency. This is not quite correct. The `genimage.bbclass`  documentation states quite clearly that this variable is intended for `.tar` images that should be unpacked and re-packed into a filesystem image (thx @ejoerns ). Instead I've now set up the dependencies explicitly.
- [x] Is the rootfs grown like we want it to.

The last one I can test, for the rest I would be glad about some feedback.

---

¹ Which is the end of the image but has nothing to do with the end of the _device_ the image is written to, where the header should be.

hnez committed:
Revert "emmc-image: remove size workaround"

This change did not do quite what we wanted it to do.
We do actually want the two rootfs partitions to be 2G in size,
not the whole image.

Getting rid of the "partition end" hack is actually something
we want, so we will re-do that in another commit but a bit differently.

This reverts commit 9b5bf5b79aeb63161435918fdb9ad4da8614b6ee.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: emmc-image: do not fill and write GPT backup header

A GPT parition table places an additional header at the _end_ of the device
it is on. Our eMMC is however 16Gb large and the image we initially deploy
is much smaller than that.
The GPT header genimage generates for us does thus not land at the end of
the eMMC and every tool looking for it there will not find it.
Writing it is thus quite useless.

Removing it means that the image now longer has to span the whole 2Gb we
previously had. It can instead end right after the file system.
To make the image end right after the file system we also need to remove the
dummy partition placed at the end of the image just to force padding of the
image from the end of the file system to the dummy partition.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: emmc-image: use existing rootfs ext4 image

We currently generate an ext4 image from a rootfs tar just for use with
genimage. This is quite redundant as we already have one we use e.g. in our
RAUC bundles. Building both is kind of wasteful.

Use the existing ext4 rootfs image.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

rohieb commented:
The seond one you could also check by inspecting the `DEPENDS` variable in the output of `bitbake -e emmc-image`.

Edit: Ah, the `do_genimage[depends] += "${GENIMAGE_ROOTFS_IMAGE}:do_image_complete"` should do the trick, as long as the `do_image_complete` task in *lxatac-core-image-base* creates the ext4 image.

hnez commented on meta-lxatac-bsp/recipes-core/images/emmc-image.bb:
```diff
@@ -12,9 +12,7 @@ inherit genimage
-GENIMAGE_ROOTFS_IMAGE_FSTYPE = "tar"
 
 do_genimage[depends] += " \
-    ${GENIMAGE_ROOTFS_IMAGE}:do_image_complete \
+    lxatac-core-image-base:do_image_complete \
 "
```
I've changed how the dependency on the rootfs image is handled because `GENIMAGE_ROOTFS_IMAGE` is no longer set.
See the updated PR description for more detail.

jluebbe replied:
Looks good.

hnez commented:
Seeing the `core` and `base` in the image name always irks me a little bit, as it is not very core-y or base-y with all the development tools.
We should define an actual minimal image in the future and derive our full image with all the tools from there.
(But that's not something we should do or discuss here)

hnez committed:
meta-lxatac-bsp: systemd: backport systemd-growfs-root.service

The systemd-growfs-root.service file is provided by later systemd versions
and, when enabled, resizes the current root filesystem to fill the
partition it resides on.
We want that feature now, so backport the service file and enable it.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez commented:
As it turns out the rootfs was not resized to match the partition as we wanted it to.
Newer `systemd` versions ship a `systemd-growfs-root.service` that can help us here, so I've backported it for now.
Hopefully we can throw away the backport very soon.
